src/ipred.rs

```diff
@@ -1461,39 +1469,51 @@ unsafe extern "C" fn cfl_ac_c_erased<BD: BitDepth, const IS_SS_HOR: bool, const
     );
 }
 
-unsafe fn pal_pred_rust<BD: BitDepth>(
-    mut dst: *mut BD::Pixel,
-    stride: ptrdiff_t,
+fn pal_pred_rust<BD: BitDepth>(
+    dst: &Rav1dPictureDataComponent,
+    mut dst_offset: usize,
     pal: &[BD::Pixel; 8],
-    mut idx: *const u8,
+    idx: &[u8],
     w: c_int,
     h: c_int,
 ) {
-    let mut y = 0;
-    while y < h {
-        let mut x = 0;
-        while x < w {
-            let i = *idx;
+    let w = w as usize;
+    let h = h as usize;
+    let idx = &idx[..w * h / 2];
+
+    let mut j = 0;
+    for _ in 0..h {
+        for x in (0..w).step_by(2) {
+            let i = idx[j];
+            j += 1;
             assert!((i & 0x88) == 0);
-            *dst.offset(x as isize) = pal[(i & 7) as usize];
-            *dst.offset(x as isize + 1) = pal[(i >> 4) as usize];
-            idx = idx.offset(1);
-            x += 2;
+            let dst = &mut *dst.slice_mut::<BD, _>((dst_offset + x.., ..2));
+            dst[0] = pal[(i & 7) as usize];
+            dst[1] = pal[(i >> 4) as usize];
         }
-        dst = dst.offset(BD::pxstride(stride));
-        y += 1;
+        dst_offset = dst_offset.wrapping_add_signed(dst.pixel_stride::<BD>());
     }
 }
 
 unsafe extern "C" fn pal_pred_c_erased<BD: BitDepth>(
-    dst: *mut DynPixel,
-    stride: ptrdiff_t,
+    dst_ptr: *mut DynPixel,
+    _stride: ptrdiff_t,
     pal: *const [DynPixel; 8],
     idx: *const u8,
     w: c_int,
     h: c_int,
+    dst: *const FFISafe<Rav1dPictureDataComponent>,
 ) {
-    pal_pred_rust::<BD>(dst.cast(), stride, &*pal.cast(), idx, w, h);
+    let dst_ptr = dst_ptr.cast::<BD::Pixel>();
+    // SAFETY: Was passed as `FFISafe::new(dst)` in `pal_pred::Fn::call`.
+    let dst = unsafe { FFISafe::get(dst) };
+    // SAFETY: Reverse of what was done in `pal_pred::Fn::call`.
+    let dst_offset = unsafe { dst_ptr.offset_from(dst.as_mut_ptr::<BD>()) } as usize;
+    // SAFETY: Undoing dyn cast in `pal_pred::Fn::call`.
+    let pal = unsafe { &*pal.cast() };
+    // SAFETY: Length sliced in `pal_pred::Fn::call`.
+    let idx = slice::from_raw_parts(idx, (w * h) as usize / 2);
```

Comment:
Why no `unsafe` block around `from_raw_parts` when you have one around the other unsafe ops?